spec/lib/second_factor/auth_manager_spec.rb

```diff
@@ -0,0 +1,180 @@
+        manager.run!(create_request, { second_factor_nonce: SecureRandom.hex }, {})
+        expect(action.called_methods).to contain_exactly(
+          :no_second_factors_enabled!
+        )
+      end
+    end
+
+    it "initiates the 2FA process and stages a challenge in secure session when there is no nonce in params" do
+      set_subfolder("/community")
+      action = create_action({ call_me_back: 4314 })
+      manager = create_manager(action)
+      request = create_request(
+        request_method: "POST",
+        path: "/abc/xyz"
+      )
+      secure_session = {}
+      expect {
+        manager.run!(request, {}, secure_session)
+      }.to raise_error(SecondFactor::AuthManager::SecondFactorRequired)
+      expect(action.called_methods).to contain_exactly(
+        :second_factor_auth_required!
+      )
+      json = secure_session["current_second_factor_auth_challenge"]
+      challenge = JSON.parse(json).deep_symbolize_keys
+      expect(challenge[:nonce]).to be_present
+      expect(challenge[:callback_method]).to eq("POST")
+      expect(challenge[:callback_path]).to eq("/abc/xyz")
+      expect(challenge[:redirect_path]).to eq("/community/")
+      expect(challenge[:allowed_methods]).to eq(manager.allowed_methods.to_a)
+      expect(challenge[:callback_params]).to eq({ call_me_back: 4314 })
+    end
+
+    it "calls the second_factor_auth_completed! method of the action if the challenge is successful and not expired" do
+      nonce, secure_session = stage_challenge(successful: true)
+
+      action = create_action
+      manager = create_manager(action)
+      request = create_request(
+        request_method: "POST",
+        path: "/abc/xyz"
+      )
+      manager.run!(request, { second_factor_nonce: nonce }, secure_session)
+      expect(action.called_methods).to contain_exactly(
+        :second_factor_auth_completed!
+      )
+    end
+
+    it "does not call the second_factor_auth_completed! method of the action if the challenge is not marked successful" do
+      nonce, secure_session = stage_challenge(successful: false)
+
+      action = create_action
+      manager = create_manager(action)
+      request = create_request(
+        request_method: "POST",
+        path: "/abc/xyz"
+      )
+      expect {
+        manager.run!(request, { second_factor_nonce: nonce }, secure_session)
+      }.to raise_error(SecondFactor::BadChallenge) do |ex|
+        expect(ex.error_translation_key).to eq("second_factor_auth.challenge_not_completed")
+      end
+      expect(action.called_methods).to be_blank
+    end
+
+    it "does not call the second_factor_auth_completed! method of the action if the challenge is expired" do
+      nonce, secure_session = stage_challenge(successful: true)
+
+      action = create_action
+      manager = create_manager(action)
+      request = create_request(
+        request_method: "POST",
+        path: "/abc/xyz"
+      )
+
+      freeze_time 6.minutes.from_now
```

Comment:
I think it'll be better to use the `SecondFactor::AuthManager::MAX_CHALLENGE_AGE` constant here so that these tests don't randomly break if we were to tweak the constant.